mario/core/CoreIO.py

```diff
@@ -649,64 +649,109 @@ def is_productive(self, method: str, data_set: str = "baseline") -> bool:
-            residuals = []  # initial point
-            error = 0.1
-            i = 0
-
-            while residuals[i] >= error:
-                M[i + 1] = np.linalg.matrix_power(z, i + 1)
-                M_cum[i + 1] = M_cum[i] + M[i + 1]
-                residuals.append(sum(sum(abs(L - M_cum[i + 1]))))
-                if residuals[i + 1] > residuals[i]:
-                    log_time(
-                        logger,
-                        "Test: non-productive system (non-convergent power series)",
-                    )
-                    _productive = False
-                    break
-                i += 1
-            else:
-                log_time(
-                    logger, "Test: productive system (non-convergent power series)"
-                )
+    def is_isard(self, scenario: str = "baseline") -> bool:
 
-        elif method == "C" and __cvxpy__:
+        """Checks whether a table is in Isard format.
+        Isard SUT tables account for trades among regions in the USE matrix
 
-            x = cp.Variable((Y.shape[0], 1))
-            s = cp.Variable((Y.shape[0], 1))
+        Parameters
+        ------------
+        scenario: str
+            defining the scenario to be checked
 
-            Z = cp.sum(s, 0)
-            obj = cp.Minimize(Z)
+        RETURN
+        -------------
 
-            const = [
-                cp.matmul(I - z, x) + s == np.ones((Y.shape[0], 1)),
-                x >= 0,
-                s >= 0,
-            ]
+        boolean
+
+                True if the dataset is isard
+                Flase if the dataset is not isard
+
+        """
+
+        if self.meta.table != "SUT":
+            raise NotImplementable("This test is implementable only on SUT tables")
+        elif len(self.get_index(_MASTER_INDEX["r"])) == 1:
+            raise NotImplementable("This test is not implementable on single-region tables")
+    
+        if scenario not in self.scenarios:
+            raise WrongInput("Acceptable data_sets are:\n{}".format(self.scenarios))
+           
+        if 'z' in self.matrices[scenario]: # this avoid to calculate z or Z in case one of them is missing, to avoid losing time
+            matrix = self.matrices[scenario][_ENUM.z]
+        else:
+            matrix = self.matrices[scenario][_ENUM.Z]
+        
+        sN = slice(None)
+        matrix = matrix.loc[(sN,_MASTER_INDEX['c'],sN),(sN,_MASTER_INDEX['a'],sN)] # extract the use side from z or Z
+        matrix = matrix.groupby(level=[_MASTER_INDEX['r']],axis=0).sum()
+        matrix = matrix.groupby(level=[_MASTER_INDEX['r']],axis=1).sum()
+
+        is_diagonal = np.all(matrix.values == np.diag(np.diagonal(matrix)))
+
+        if is_diagonal:
+            log_time(logger, "Test: table is not in Isard format")
+
+            return False
+
+        else:
+            log_time(logger, "Test: table is in Isard format")
+
+            return True
+
+    def is_chenerymoses(self, scenario: str = "baseline") -> bool:
```

Comment:
is it correct to say if a table is ISARD is not chenerymoses, and if a table is not isard is chenerymoses?

if yes, we might avoid one of the two functions.